addons/delivery/models/stock_picking.py

```diff
@@ -187,6 +187,19 @@ def _add_delivery_cost_to_so(self):
         sale_order = self.sale_id
         if sale_order.invoice_shipping_on_delivery:
             sale_order._create_delivery_line(self.carrier_id, self.carrier_price)
+            delivery_line = sale_order.order_line.filtered(lambda line: line.is_delivery and line.price_unit == 0)
+            delivery_line.qty_delivered = 1
+        else:
+            delivery_line = sale_order.order_line.filtered(lambda line: line.is_delivery and line.price_unit == 0)
+            if delivery_line:
+                res = self.carrier_id.rate_shipment(sale_order)
+                if res['success']:
+                    delivery_line.qty_delivered = 1  # to trigger the invoice status
+                    delivery_line.price_unit = res['price']
```

Comment:
price_unit += res['price'] will make the backorder work (example: you have to deliver 3, you deliver 2 then backorder 1) moreover, don't call rate, when you send a shippment you get the price back in 'exact_price' which is what you want to add on your saleorder. (because by rating the sale order again, the price will not match what you paid (changing the package, send more, send less,...) And it does one more request to an external API, that takes time

Reply:
This should do the trick: https://github.com/odoo-dev/odoo/commit/90086c8e424cb15c85c0dc4385e9c1c4de9ad03b